kotlin/samples/containers/app-porchbox/src/test/java/com/squareup/sample/porchbox/PorchWorkflowTest.kt

```diff
@@ -0,0 +1,116 @@
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun signInWithEmailAndPassword(
+      email: String,
+      pass: String,
+      listener: OnCompleteListener<AuthResult>
+    ): Task<AuthResult> {
+      val authTaskResult = Mockito.mock(Task::class.java)
+      val authResult = Mockito.mock(AuthResult::class.java)
+      Mockito.`when`(authTaskResult.result).thenReturn(authResult)
+      Mockito.`when`(authTaskResult.isSuccessful).thenReturn(true)
+      Mockito.`when`(authResult.user).thenReturn(currentUser)
+      listener.onComplete(authTaskResult as Task<AuthResult>)
+      return authTaskResult
+    }
+  }
+
+  object FakeRealtimeDb: RealtimeDB {
+    private val testInbox = listOf(
+        Mockito.mock(DocumentSnapshot::class.java).also {
+          Mockito.`when`(it.getString("name")).thenReturn("Lysol Wipes")
+          Mockito.`when`(it.getString("photoUri")).thenReturn("https://picsum.photos/id/1021/200/200.jpg")
+        },
+        Mockito.mock(DocumentSnapshot::class.java).also {
+          Mockito.`when`(it.getString("name")).thenReturn("Groceries")
+          Mockito.`when`(it.getString("photoUri")).thenReturn("https://picsum.photos/id/1021/200/200.jpg")
+        },
+        Mockito.mock(DocumentSnapshot::class.java).also {
+          Mockito.`when`(it.getString("name")).thenReturn("Shopvac")
+          Mockito.`when`(it.getString("photoUri")).thenReturn("https://picsum.photos/id/1021/200/200.jpg")
+        },
+        Mockito.mock(DocumentSnapshot::class.java).also {
+          Mockito.`when`(it.getString("name")).thenReturn("Book")
+          Mockito.`when`(it.getString("photoUri")).thenReturn("https://picsum.photos/id/1021/200/200.jpg")
+        }
+    )
+
+    override fun query(
+      path: String,
+      field: String?,
+      subset: List<String?>,
+      listener: OnSuccessListener<QuerySnapshot>
+    ): Task<QuerySnapshot> {
+      val queryTaskResult = Mockito.mock(Task::class.java)
+      val snapshot = Mockito.mock(QuerySnapshot::class.java)
+      Mockito.`when`(queryTaskResult.result).thenReturn(snapshot)
+      Mockito.`when`(queryTaskResult.isSuccessful).thenReturn(true)
+      Mockito.`when`(snapshot.documents).thenReturn(
+          testInbox
+      )
+      listener.onSuccess(snapshot)
+      @Suppress("UNCHECKED_CAST")
+      return queryTaskResult as Task<QuerySnapshot>
+    }
+  }
+
+  @Test fun `starts in Login`() {
+    val workflow = PorchWorkflow(FakeAuthSuccess, FakeRealtimeDb);
+
+    workflow.testFromStart(Props) {
+      assertThat(awaitNextRendering().beneathModals is LoginRendering).isTrue()
+    }
+  }
+
+  @Test fun `test login loading`() {
+    PorchWorkflow(FakeAuthSuccess, FakeRealtimeDb).testFromState(
+        Props,
+        LoggingIn(BuildConfig.CLIENT_EMAIL, BuildConfig.CLIENT_PASS)) {
+      assertThat(awaitNextRendering().beneathModals).isEqualTo(LoggingIn(BuildConfig.CLIENT_EMAIL, BuildConfig.CLIENT_PASS))
```

Comment:
@zach-klippenstein this test is failing but I am having trouble understanding why.  Yes, I know that my fakeAuth and fakeDb will return immediately, but still there should be a Loading rendering *first* before the data is loaded (regardless of how quickly the Action is handled).  And yet, the result of awaitNextRendering() is always the rendering of the full data set Master Detail?  What am I missing?